src/NLog/Internal/FilePathLayout.cs

```diff
@@ -0,0 +1,294 @@
+            {
+                return rendered;
+            }
+#if !SILVERLIGHT
+            if (_filePathKind == FilePathKind.Relative)
+            {
+                return Path.Combine(_baseDir, rendered);
+                //use basedir, faster than Path.GetFullPath
+            }
+#endif
+            //unknown, use slow method
+            return Path.GetFullPath(rendered);
+
+        }
+
+        #endregion
+
+
+        /// <summary>
+        /// Is this (templated/invalid) path an absolute, relative or unknown?
+        /// </summary>
+        internal static FilePathKind DetectFilePathKind(Layout pathLayout)
+        {
+            var simpleLayout = pathLayout as SimpleLayout;
+            if (simpleLayout == null)
+            {
+                return FilePathKind.Unknown;
+            }
+
+            return DetectFilePathKind(simpleLayout);
+        }
+        /// <summary>
+        /// Is this (templated/invalid) path an absolute, relative or unknown?
+        /// </summary>
+        private static FilePathKind DetectFilePathKind(SimpleLayout pathLayout)
+        {
+            var isFixedText = pathLayout.IsFixedText;
+
+            //nb: ${basedir} has already been rewritten in the SimpleLayout.compile
+            var path = isFixedText ? pathLayout.FixedText : pathLayout.Text;
+
+            if (path != null)
+            {
+                path = path.TrimStart();
+
+                int length = path.Length;
+                if (length >= 1)
+                {
+                    var firstChar = path[0];
+                    if (firstChar == Path.DirectorySeparatorChar || firstChar == Path.AltDirectorySeparatorChar)
+                        return FilePathKind.Absolute;
+                    if (firstChar == '.') //. and ..
+                    {
+                        return FilePathKind.Relative;
+                    }
+
+
+                    if (length >= 2)
+                    {
+                        var secondChar = path[1];
+                        if (secondChar == Path.VolumeSeparatorChar)
+                            return FilePathKind.Absolute;
+
+                    }
+                    if (!isFixedText && path.StartsWith("${", StringComparison.OrdinalIgnoreCase))
+                    {
+                        //if first part is a layout, then unknown
+                        return FilePathKind.Unknown;
+                    }
+
+
+                    //not a layout renderer, but text
+                    return FilePathKind.Relative;
+                }
+            }
+            return FilePathKind.Unknown;
+        }
+
+        private static string CleanupInvalidFileName(string fileName)
+        {
+#if !SILVERLIGHT
+            if (StringHelpers.IsNullOrWhiteSpace(fileName))
+            {
+                return fileName;
+            }
+
+
+            var lastDirSeparator = fileName.LastIndexOfAny(DirectorySeparatorChars);
+
+            var fileName1 = fileName.Substring(lastDirSeparator + 1);
```

Comment:
This code would be cleaner if we referred to `filePath` and `fileName`, rather than `fileName` and `fileName1`.